File: strava_sync.ex
Mix.install([
  {:req, "~> 0.2"},
  {:timex, "~> 3.7.9"}
])

strava_client_id = System.get_env("STRAVA_CLIENT_ID")
strava_refresh_token = System.get_env("STRAVA_REFRESH_TOKEN")
strava_client_secret = System.get_env("STRAVA_CLIENT_SECRET")

refresh_token_url =
  "https://www.strava.com/oauth/token?" <>
    "grant_type=refresh_token&refresh_token=#{strava_refresh_token}&" <>
    "client_id=#{strava_client_id}&client_secret=#{strava_client_secret}"

%{
  body: %{
    "access_token" => access_token,
    "refresh_token" => refresh_token
  }
} = Req.post!(refresh_token_url)

if System.get_env("DBEUG") == "1" do
  IO.puts(
    "\nAccess token: #{access_token}\nRefresh token: #{refresh_token}\n\n" <>
      "refresh token url: #{refresh_token_url}"
  )
end

after_unix = Date.utc_today() |> Timex.shift(months: -1) |> Timex.to_unix()

{:ok, %{body: activities}} =
  Req.get("https://www.strava.com/api/v3/athlete/activities?after=#{after_unix}&per_page=200",
    headers: [{"Authorization", "Bearer #{access_token}"}]
  )

summary_acc = %{
  total_distance: 0,
  total_time: 0,
  avg_pace: 0,
  total_elevation_gain: 0
}

{summary, number_of_runs} =
  activities
  |> Enum.filter(&(&1["type"] == "Run"))
  |> Enum.reduce({summary_acc, 0}, fn %{
                                        "distance" => run_distance,
                                        "elapsed_time" => run_time,
                                        "average_speed" => run_pace,
                                        "total_elevation_gain" => run_ascent
                                      },
                                      {%{
                                         total_distance: total_distance,
                                         total_time: total_time,
                                         avg_pace: avg_pace,
                                         total_elevation_gain: total_elevation_gain
                                       }, index} ->
    {
      %{
        total_distance: total_distance + run_distance,
        total_time: total_time + run_time,
        avg_pace: avg_pace + run_pace,
        total_elevation_gain: total_elevation_gain + run_ascent
      },
      index + 1
    }
  end)

defmodule Convert do
  @minute 60
  @hour @minute * 60
  @day @hour * 24
  @week @day * 7
  @divisor [@week, @day, @hour, @minute, 1]

  def sec_to_str(sec) do
    {_, [s, m, h, d, w]} =
      Enum.reduce(@divisor, {sec, []}, fn divisor, {n, acc} ->
        {rem(n, divisor), [div(n, divisor) | acc]}
      end)

    ["#{w} wk", "#{d} d", "#{h} hr", "#{m} min", "#{s} sec"]
    |> Enum.reject(fn str -> String.starts_with?(str, "0") end)
    |> Enum.join(", ")
  end
end

new_readme =
  "README.md.eex"
  |> File.read!()
  |> EEx.eval_string(
    number_of_runs: number_of_runs,
    total_elevation_gain: :erlang.float_to_binary(summary.total_elevation_gain, decimals: 2),
    total_time: Convert.sec_to_str(summary.total_time),
    total_distance:
      summary.total_distance |> Kernel./(1000) |> :erlang.float_to_binary(decimals: 2),
    from_date: after_unix |> DateTime.from_unix!() |> DateTime.to_date(),
    to_date: DateTime.utc_now() |> DateTime.to_date()
  )

File.write!("README.md", new_readme)

File: strava_sync/src/strava_sync.erl
-module(strava_sync).

-export([main/1, refresh_token_url/0, activities_url/2, make_request/5,
         fetch_all_activities/3, twelve_months_ago/0, runs_summary/2, empty_summary/0, add_run/2]).

-type args() :: list().
-type epoch() :: calendar:time().
-type method() :: hackney:method().
-type url() :: hackney:url().
-type headers() :: hackney:headers().
-type body() :: hackney:body().
-type options() :: hackney:options().
-type summary() :: #{list() => number() | list()}.

%% escript Entry point
-spec main(args()) -> no_return().
main(_Args) ->
  application:ensure_all_started(hackney),
  RefreshTokenUrl = refresh_token_url(),
  #{<<"access_token">> := AccessToken} = make_request(post, RefreshTokenUrl, [], <<>>, []),
  AfterUnix = integer_to_list(twelve_months_ago()),
  Activities = fetch_all_activities(AfterUnix, AccessToken, 1),
  RunsSummary = runs_summary(Activities, empty_summary()),
  Template = bbmustache:parse_file(<<"../README.mustache">>),
  file:write_file("../README.md", bbmustache:compile(Template, RunsSummary)),
  io:format("Summary: ~p~nTwelve months ago: ~p~n", [RunsSummary, twelve_months_ago()]),
  erlang:halt(0).

-spec refresh_token_url() -> list().
refresh_token_url() ->
  StravaClientId = os:getenv("STRAVA_CLIENT_ID"),
  StravaRefreshToken = os:getenv("STRAVA_REFRESH_TOKEN"),
  StravaClientSecret = os:getenv("STRAVA_CLIENT_SECRET"),
  "https://www.strava.com/oauth/token?"
  ++ "grant_type=refresh_token&refresh_token="
  ++ StravaRefreshToken
  ++ "&"
  ++ "client_id="
  ++ StravaClientId
  ++ "&client_secret="
  ++ StravaClientSecret.

-spec activities_url(epoch(), integer()) -> list().
activities_url(AfterUnix, Page) ->
  "https://www.strava.com/api/v3/athlete/activities?after="
  ++ AfterUnix
  ++ "&page="
  ++ integer_to_list(Page)
  ++ "&per_page=200".

-spec make_request(method(), url(), headers(), body(), options()) -> [map()] | map().
make_request(Method, Url, Headers, Payload, Options) ->
  {ok, _S, _H, Ref} = hackney:request(Method, Url, Headers, Payload, Options),
  {ok, Body} = hackney:body(Ref),
  jsx:decode(Body, [{return_maps, true}]).

%% Recursive activity fetch
-spec fetch_all_activities(epoch(), binary(), integer()) -> [map()].
fetch_all_activities(AfterUnix, AccessToken, Page) ->
  ActivitiesUrl = activities_url(AfterUnix, Page),
  Headers = [{<<"Authorization">>, <<"Bearer ", AccessToken/binary>>}],
  Activities = make_request(get, ActivitiesUrl, Headers, <<>>, []),
  case length(Activities) of
    N when N < 200 ->
      Activities;
    _ ->
      Activities ++ fetch_all_activities(AfterUnix, AccessToken, Page + 1)
  end.

-spec twelve_months_ago() -> epoch().
twelve_months_ago() ->
  TwelveMonthsAgo =
    edate:shift(
      edate:today(), -12, month),
  EpochSeconds = calendar:datetime_to_gregorian_seconds({{1970, 1, 1}, {0, 0, 0}}),
  calendar:datetime_to_gregorian_seconds({TwelveMonthsAgo, {0, 0, 0}}) - EpochSeconds.

-spec runs_summary([map()], summary()) -> summary().
runs_summary([],
             Acc =
               #{"total_time" := TotalTime,
                 "total_distance" := TotalDistance,
                 "total_elevation_gain" := TotalElevationGain}) ->
  Acc#{"total_time" => convert_seconds:convert(TotalTime),
       "total_distance" => float_to_binary(TotalDistance / 1000, [{decimals, 2}]),
       "total_elevation_gain" => float_to_binary(TotalElevationGain, [{decimals, 2}])};
runs_summary([#{<<"type">> := <<"Run">>} = Run | Rest], Acc) ->
  runs_summary(Rest, add_run(Acc, Run));
runs_summary([_ | Rest], Acc) ->
  runs_summary(Rest, Acc).

-spec empty_summary() -> summary().
empty_summary() ->
  #{"total_distance" => 0,
    "total_time" => 0,
    "avg_pace" => 0,
    "total_elevation_gain" => 0,
    "number_of_runs" => 0,
    "from_date" =>
      edate:date_to_string(
        edate:shift(
          edate:today(), -12, month)),
    "to_date" =>
      edate:date_to_string(
        edate:today())}.

-spec add_run(summary(), map()) -> summary().
add_run(#{"total_distance" := TotalDistance,
          "total_time" := TotalTime,
          "avg_pace" := AvgPace,
          "total_elevation_gain" := TotalElevationGain,
          "number_of_runs" := NumberOfRuns} =
          Summary,
        #{<<"distance">> := RunDistance,
          <<"elapsed_time">> := RunTime,
          <<"average_speed">> := RunPace,
          <<"total_elevation_gain">> := RunAscent}) ->
  Summary#{"total_distance" => TotalDistance + RunDistance,
           "total_time" => TotalTime + RunTime,
           "avg_pace" => AvgPace + RunPace,
           "total_elevation_gain" => TotalElevationGain + RunAscent,
           "number_of_runs" => NumberOfRuns + 1}.

-ifdef(TEST).

-include_lib("eunit/include/eunit.hrl").

summary_test_() ->
  {ok, Activities} = file:read_file("activities.json"),
  ActivitiesDecoded = jsx:decode(Activities, [{return_maps, true}]),
  RunsSummary = runs_summary(ActivitiesDecoded, empty_summary()),

  [?_assertEqual(#{"avg_pace" => 492.7909999999999,
                   "from_date" =>
                     edate:date_to_string(
                       edate:shift(
                         edate:today(), -12, month)),
                   "number_of_runs" => 168,
                   "to_date" =>
                     edate:date_to_string(
                       edate:today()),
                   "total_distance" => <<"1054.88">>,
                   "total_elevation_gain" => <<"15813.00">>},
                 maps:remove("total_time", RunsSummary))].

% #TODO total_time
-endif.
